src/mgt/clMgtDatabase.cxx

```diff
@@ -144,19 +152,121 @@ namespace SAFplus
         return rc;
       }
 
-    Mgt::Msg::MsgMgtDb dbValue;
     std::string strVal;
-    strVal.append(cvalue, dataSize);
+    strVal.assign(cvalue, dataSize);
 
     // De-marshall data
     dbValue.ParseFromString(strVal);
+    SAFplusHeapFree(cvalue);
+    return rc;
+  }
+
+  ClRcT MgtDatabase::getRecord(const std::string &key, std::string &value)
+  {
+    ClRcT rc = CL_OK;
+
+    Mgt::Msg::MsgMgtDb dbValue;
+    rc = getRecordDbValue(key, dbValue);
+    if (CL_OK != rc)
+      {
+        return rc;
+      }
+
+    ClUint32T hashKey = getHashKeyFn(key.c_str());
     value.assign(dbValue.value());
 
     logInfo("MGT", "DBR", "Record [0x%x]: [%s] -> [%s]", hashKey, key.c_str(), value.c_str());
-    SAFplusHeapFree(cvalue);
     return rc;
   }
 
+  ClRcT MgtDatabase::insertByDbValue(const std::string &key,
+      const Mgt::Msg::MsgMgtDb &dbValue)
+  {
+    ClRcT rc = CL_OK;
+
+    ClUint32T hashKey = getHashKeyFn(key.c_str());
+    std::string strVal;
+    // Marshall data
+    dbValue.SerializeToString(&strVal);
+    /*
+     * Insert into data table
+     */
+    rc = clDbalRecordInsert(mDbDataHdl, (ClDBKeyT) &hashKey, sizeof(hashKey), (ClDBRecordT) strVal.c_str(), strVal.size());
+    return rc;
+  }
+
+  void MgtDatabase::insertToParentRecord(const std::string &key)
+  {
+    ClRcT rc = CL_OK;
+    std::string childName;
+    std::string parentKey;
+    std::size_t found = key.find_last_of("/");
+    if (found == std::string::npos || found == 0)
+      {
+        return;
+      }
+
+    if (key[key.length() - 1] == ']')
+      {
+        found = key.find_last_of("[");
+        if (found != std::string::npos)
+          {
+            childName = key.substr(found);
+            parentKey = key.substr(0, found);
+          }
+      }
+    else
+      {
+        childName = key.substr(found + 1);
+        parentKey = key.substr(0, found);
+      }
+
+    if (std::find(listKey.begin(), listKey.end(), parentKey)!= listKey.end())
+      {
+        Mgt::Msg::MsgMgtDb dbValue;
+        rc = getRecordDbValue(parentKey, dbValue);
```

Comment:
This is less performance with read and write into db (file access) in loop listKey